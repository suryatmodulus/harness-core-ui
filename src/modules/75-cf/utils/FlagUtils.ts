/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { Feature, GitDetails, PatchFeatureQueryParams, usePatchFeature } from 'services/cf'

export interface FlagPatchParams {
  accountIdentifier: string
  orgIdentifier: string
  projectIdentifier: string
  environmentIdentifier: string
}

const makeInstruction = (kind: string, variation: string, targetIdentifiers: string[], gitDetails?: GitDetails) => {
  const instructions = {
    instructions: [
      {
        kind,
        parameters: {
          variation,
          targets: targetIdentifiers
        }
      }
    ]
  }

  return gitDetails ? { ...instructions, gitDetails } : instructions
}

const makePatchHook =
  (patchKind: string) =>
  ({ accountIdentifier, orgIdentifier, projectIdentifier, environmentIdentifier }: FlagPatchParams) => {
    const { mutate } = usePatchFeature({
      identifier: '',
      queryParams: {
        account: accountIdentifier,
        accountIdentifier,
        org: orgIdentifier,
        project: projectIdentifier,
        environment: environmentIdentifier
      } as PatchFeatureQueryParams
    })

    return (featureFlag: Feature, variation: any, targetIdentifiers: string[], gitDetails?: GitDetails) => {
      const body = makeInstruction(patchKind, variation, targetIdentifiers, gitDetails)

      return mutate(body, { pathParams: { identifier: featureFlag.identifier } })
    }
  }

export const useServeFeatureFlagVariationToTargets = makePatchHook('addTargetsToVariationTargetMap')

export const useRemoveFeatureFlagVariationFromTargets = makePatchHook('removeTargetsToVariationTargetMap')
