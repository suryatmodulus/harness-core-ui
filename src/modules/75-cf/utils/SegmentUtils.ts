/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { PatchSegmentQueryParams, usePatchSegment } from 'services/cf'

export interface TargetSegmentParams {
  accountIdentifier: string
  orgIdentifier: string
  environmentIdentifier: string
  projectIdentifier: string
}

const makeInstructions = (kind: string, targetIdentifiers: string[]) => ({
  instructions: [
    {
      kind,
      parameters: {
        targets: targetIdentifiers
      }
    }
  ]
})

const makeTargetsToSegmentHook =
  (patchKind: string) =>
  ({ accountIdentifier, orgIdentifier, environmentIdentifier, projectIdentifier }: TargetSegmentParams) => {
    const { mutate } = usePatchSegment({
      identifier: '',
      queryParams: {
        account: accountIdentifier,
        accountIdentifier,
        org: orgIdentifier,
        project: projectIdentifier,
        environment: environmentIdentifier
      } as PatchSegmentQueryParams
    })

    return (segmentIdentifier: string, targetIdentifiers: string[]) => {
      const body = makeInstructions(patchKind, targetIdentifiers)
      return mutate(body, { pathParams: { identifier: segmentIdentifier } })
    }
  }

/** Add targets to an include segment list */
export const useAddTargetsToIncludeList = makeTargetsToSegmentHook('addToIncludeList')

/** Remove targets to an include segment list */
export const useRemoveTargetsFromIncludeList = makeTargetsToSegmentHook('removeFromIncludeList')

/** Add targets to an exclude segment list */
export const useAddTargetsToExcludeList = makeTargetsToSegmentHook('addToExcludeList')

/** Remove targets to an exclude segment list */
export const useRemoveTargetsFromExcludeList = makeTargetsToSegmentHook('removeFromExcludeList')
