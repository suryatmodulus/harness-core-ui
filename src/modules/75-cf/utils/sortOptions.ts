/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export interface Option {
  label: string
  value: string
}

export const sortOptions = (options: Option[]): Option[] =>
  options.sort(({ label: a }, { label: b }) => a.localeCompare(b))
