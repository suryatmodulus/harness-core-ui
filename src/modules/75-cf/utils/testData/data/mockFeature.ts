/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { Feature } from 'services/cf'

const mockFeature: Feature = {
  archived: false,
  createdAt: 1635243729074,
  defaultOffVariation: 'false',
  defaultOnVariation: 'true',
  description: '',
  envProperties: {
    defaultServe: { variation: 'false' },
    environment: 'testnonprod',
    modifiedAt: 1635333973373,
    offVariation: 'false',
    rules: [],
    state: 'on',
    variationMap: [
      {
        targetSegments: ['test_target_group'],
        targets: [{ identifier: 'another_target', name: 'another target' }],
        variation: 'true'
      }
    ],
    version: 56
  },
  evaluation: 'false',
  evaluationIdentifier: 'false',
  identifier: 'new_flag',
  kind: 'boolean',
  modifiedAt: 1635333973371,
  name: 'new flag ',
  owner: ['chris.blakely@harness.io'],
  permanent: false,
  prerequisites: [],
  project: 'chrisgit2',
  results: undefined,
  status: { lastAccess: -6795364578871, status: 'never-requested' },
  tags: [],
  variations: [
    { identifier: 'true', name: 'True', value: 'true' },
    { identifier: 'false', name: 'False', value: 'false' }
  ]
}

export default mockFeature
