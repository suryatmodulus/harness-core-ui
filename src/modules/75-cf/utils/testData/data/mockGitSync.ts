/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import * as yup from 'yup'
import type { UseGitSync } from '@cf/hooks/useGitSync'

const mockGitSync: UseGitSync = {
  gitRepoDetails: {
    autoCommit: false,
    branch: 'main',
    filePath: '/flags.yaml',
    objectId: '',
    repoIdentifier: 'harnesstest',
    rootFolder: '/.harness/'
  },
  isAutoCommitEnabled: false,
  isGitSyncEnabled: true,
  gitSyncLoading: false,
  handleGitPause: jest.fn(),
  isGitSyncActionsEnabled: true,
  isGitSyncPaused: false,
  apiError: '',
  handleError: jest.fn(),
  handleAutoCommit: jest.fn(),
  getGitSyncFormMeta: jest.fn(() => ({
    gitSyncInitialValues: {
      gitDetails: {
        commitMsg: '',
        branch: 'main',
        filePath: '/flags.yaml',
        repoIdentifier: 'harnesstest',
        rootFolder: '/.harness/'
      },
      autoCommit: false
    },
    gitSyncValidationSchema: yup.object().shape({
      commitMsg: yup.string()
    })
  }))
}

export default mockGitSync
