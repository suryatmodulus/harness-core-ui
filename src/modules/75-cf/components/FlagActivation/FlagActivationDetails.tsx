/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { ReactElement } from 'react'
import { useParams, Link } from 'react-router-dom'
import moment from 'moment'
import { Color, Layout, Text, FlexExpander, Container, Heading } from '@wings-software/uicore'
import routes from '@common/RouteDefinitions'
import { useStrings } from 'framework/strings'
import type { StringKeys } from 'framework/strings'
import { TagsViewer } from '@common/components/TagsViewer/TagsViewer'
import StringWithTooltip from '@common/components/StringWithTooltip/StringWithTooltip'
import {
  DeleteFeatureFlagQueryParams,
  Feature,
  PatchFeatureQueryParams,
  useDeleteFeatureFlag,
  usePatchFeature,
  Variation
} from 'services/cf'
import { VariationWithIcon } from '@cf/components/VariationWithIcon/VariationWithIcon'
import { useQueryParams } from '@common/hooks'
import { ResourceType } from '@rbac/interfaces/ResourceType'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import { useFeatureFlagTypeToStringMapping } from '@cf/utils/CFUtils'

import type { UseGitSync } from '@cf/hooks/useGitSync'
import { FlagTypeVariations } from '../CreateFlagDialog/FlagDialogUtils'
import { VariationTypeIcon } from '../VariationTypeIcon/VariationTypeIcon'
import { IdentifierText } from '../IdentifierText/IdentifierText'
import { EditVariationsModal } from '../EditVariationsModal/EditVariationsModal'
import { FlagPrerequisites } from './FlagPrerequisites'
import FlagDetailsOptionsMenuButton from '../FlagDetailsOptionsMenuButton/FlagDetailsOptionsMenuButton'
import css from './FlagActivationDetails.module.scss'

interface FlagActivationDetailsProps {
  featureFlag: Feature
  gitSync: UseGitSync
  gitSyncActionsComponent?: ReactElement
  refetchFlag: () => void
}

const VariationItem: React.FC<{ variation: Variation; index: number }> = ({ variation, index }) => {
  return (
    <Layout.Horizontal className={css.variationItem} spacing="xsmall" style={{ alignItems: 'center' }}>
      <VariationWithIcon variation={variation} index={index} />
    </Layout.Horizontal>
  )
}

const VariationsList: React.FC<{ featureFlag: Feature; onEditSuccess: () => void; gitSync: UseGitSync }> = ({
  featureFlag,
  onEditSuccess,
  gitSync
}) => {
  const { orgIdentifier, accountId, projectIdentifier } = useParams<Record<string, string>>()
  const isFlagTypeBoolean = featureFlag.kind === FlagTypeVariations.booleanFlag
  const { variations } = featureFlag
  const { getString } = useStrings()
  const typeToStringMapping = useFeatureFlagTypeToStringMapping()

  return (
    <Layout.Vertical padding="large" margin={{ top: 'large' }} className={css.module}>
      <Layout.Horizontal flex={{ align: 'center-center' }} margin={{ bottom: 'medium' }}>
        <Text style={{ color: '#1C1C28', fontWeight: 600, fontSize: '14px', lineHeight: '22px' }}>
          <StringWithTooltip stringId="cf.shared.variations" tooltipId="ff_ffVariations_heading" />
        </Text>
        <FlexExpander />
        <EditVariationsModal
          gitSync={gitSync}
          accountId={accountId}
          orgIdentifier={orgIdentifier}
          projectIdentifier={projectIdentifier}
          feature={featureFlag}
          permission={{
            resource: { resourceType: ResourceType.FEATUREFLAG },
            permission: PermissionIdentifier.EDIT_FF_FEATUREFLAG
          }}
          onSuccess={onEditSuccess}
          minimal
          intent="primary"
          icon="edit"
          disabled={featureFlag.archived}
        />
      </Layout.Horizontal>

      <Layout.Vertical className={css.variationsList}>
        <Text
          border={{ bottom: true, color: Color.GREY_300 }}
          padding={{ bottom: 'small' }}
          flex
          style={{ fontSize: '14px', lineHeight: '20px' }}
        >
          <VariationTypeIcon style={{ transform: 'translateY(1px)' }} multivariate={!isFlagTypeBoolean} />
          {isFlagTypeBoolean ? getString('cf.boolean') : getString('cf.multivariate')}
          {!isFlagTypeBoolean && (
            <Text color={Color.GREY_400} padding={{ left: 'xsmall' }}>
              ({typeToStringMapping[featureFlag.kind] || ''})
            </Text>
          )}
          <FlexExpander />
          <Text color={Color.GREY_400}>
            {getString('cf.featureFlagDetail.variationCount', { count: variations.length })}
          </Text>
        </Text>

        {featureFlag.variations.map((variation, index) => (
          <VariationItem key={variation.identifier} variation={variation} index={index} />
        ))}
      </Layout.Vertical>
    </Layout.Vertical>
  )
}

const FlagActivationDetails: React.FC<FlagActivationDetailsProps> = props => {
  const urlQuery: Record<string, string> = useQueryParams()
  const { featureFlag, refetchFlag, gitSyncActionsComponent, gitSync } = props
  const { getString } = useStrings()
  const { orgIdentifier, accountId, projectIdentifier } = useParams<Record<string, string>>()
  const featureFlagListURL =
    routes.toCFFeatureFlags({
      projectIdentifier: projectIdentifier,
      orgIdentifier: orgIdentifier,
      accountId
    }) + `${urlQuery?.activeEnvironment ? `?activeEnvironment=${urlQuery.activeEnvironment}` : ''}`
  const { mutate: submitPatch } = usePatchFeature({
    identifier: featureFlag.identifier as string,
    queryParams: {
      project: featureFlag.project as string,
      environment: featureFlag.envProperties?.environment as string,
      account: accountId,
      accountIdentifier: accountId,
      org: orgIdentifier
    } as PatchFeatureQueryParams
  })

  const { mutate: deleteFeatureFlag } = useDeleteFeatureFlag({
    queryParams: {
      project: projectIdentifier as string,
      account: accountId,
      accountIdentifier: accountId,
      org: orgIdentifier
    } as DeleteFeatureFlagQueryParams
  })

  const queryParams = {
    account: accountId,
    accountIdentifier: accountId,
    org: orgIdentifier,
    project: projectIdentifier,
    environment: featureFlag.envProperties?.environment as string
  } as PatchFeatureQueryParams

  const renderTime = (time: number, langString: StringKeys, style?: React.CSSProperties): React.ReactNode => (
    <Text
      style={{
        fontWeight: 500,
        lineHeight: '14px',
        fontSize: '10px',
        color: '#555770',
        letterSpacing: '0.2px',
        ...style
      }}
    >
      {getString(langString, {
        date: moment(time).format('MMMM D, YYYY hh:mm A')
      })}
    </Text>
  )

  return (
    <>
      <Layout.Horizontal className={css.breadcrumb}>
        <Link style={{ color: '#0092E4', fontSize: '12px' }} to={featureFlagListURL}>
          {getString('flag')}
        </Link>

        <span style={{ display: 'inline-block', paddingLeft: 'var(--spacing-xsmall)' }}>/</span>
        <FlexExpander />
        <FlagDetailsOptionsMenuButton
          featureFlag={featureFlag}
          gitSync={gitSync}
          queryParams={queryParams}
          submitPatch={submitPatch}
          deleteFeatureFlag={deleteFeatureFlag}
          refetchFlag={refetchFlag}
        />
      </Layout.Horizontal>

      {gitSyncActionsComponent && (
        <Container flex={{ justifyContent: 'space-between', alignItems: 'center' }} width={230}>
          {gitSyncActionsComponent}
        </Container>
      )}

      <Container>
        <Heading
          style={{
            fontWeight: 600,
            fontSize: '16px',
            lineHeight: '22px',
            color: '#1C1C28',
            marginBottom: 'var(--spacing-small)'
          }}
        >
          {featureFlag.name}
          {featureFlag.archived && (
            <Text inline color={Color.GREY_400} padding={{ left: 'xsmall' }} font={{ size: 'small' }}>
              ({getString('cf.shared.archived')})
            </Text>
          )}
        </Heading>
        {featureFlag.description && (
          <Text margin={{ bottom: 'small' }} style={{ fontSize: '13px', lineHeight: '20px', color: '#22222A' }}>
            {featureFlag.description}
          </Text>
        )}

        <IdentifierText identifier={featureFlag.identifier} allowCopy />

        {/* Disable tags for now (backend does not support them) */}
        {false && !!featureFlag.tags?.length && (
          <Container className={css.tagsFlagActivationDetails}>
            <TagsViewer
              tags={featureFlag.tags?.map(({ value }) => value as string)}
              style={{ backgroundColor: '#D9DAE6', fontSize: '12px', lineHeight: '16px', color: '#22222A' }}
            />
          </Container>
        )}

        <Layout.Vertical margin={{ top: 'medium', bottom: 'xlarge' }}>
          {renderTime(featureFlag.createdAt, 'cf.featureFlags.createdDate')}
          {renderTime(featureFlag.modifiedAt, 'cf.featureFlags.modifiedDate', { paddingTop: 'var(--spacing-xsmall)' })}
        </Layout.Vertical>

        <VariationsList
          featureFlag={featureFlag}
          gitSync={gitSync}
          onEditSuccess={() => {
            refetchFlag()
          }}
        />

        <FlagPrerequisites featureFlag={featureFlag} refetchFlag={refetchFlag} gitSync={gitSync} />
      </Container>
    </>
  )
}

export default FlagActivationDetails
