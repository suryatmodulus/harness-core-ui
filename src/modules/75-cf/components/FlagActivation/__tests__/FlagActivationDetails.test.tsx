/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, RenderResult, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { TestWrapper } from '@common/utils/testUtils'
import * as cfServiceMock from 'services/cf'

import mockFeature from '@cf/utils/testData/data/mockFeature'
import mockGitSync from '@cf/utils/testData/data/mockGitSync'
import FlagActivationDetails from '../FlagActivationDetails'

jest.mock('services/cf')

jest.mock('@cf/hooks/useEnvironmentSelectV2', () => ({
  useEnvironmentSelectV2: jest.fn().mockReturnValue({
    data: [],
    loading: false,
    error: undefined,
    refetch: jest.fn(),
    EnvironmentSelect: function EnvironmentSelect() {
      return <div />
    }
  })
}))

const renderComponent = (): RenderResult => {
  return render(
    <TestWrapper
      path="/account/:accountId/cf/orgs/:orgIdentifier/projects/:projectIdentifier/feature-flags"
      pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
    >
      <FlagActivationDetails gitSync={mockGitSync} featureFlag={mockFeature} refetchFlag={jest.fn()} />
    </TestWrapper>
  )
}

describe('FlagActivationDetails', () => {
  beforeEach(() => {
    jest.clearAllMocks()

    jest
      .spyOn(cfServiceMock, 'useGetAllFeatures')
      .mockReturnValue({ data: [], loading: false, mutate: jest.fn() } as any)
    jest
      .spyOn(cfServiceMock, 'useGetTargetsAndSegmentsInfo')
      .mockReturnValue({ data: [], loading: false, mutate: jest.fn() } as any)
    jest
      .spyOn(cfServiceMock, 'useGetAllSegments')
      .mockReturnValue({ data: [], loading: false, mutate: jest.fn() } as any)
    jest
      .spyOn(cfServiceMock, 'useGetAllTargetAttributes')
      .mockReturnValue({ data: [], loading: false, mutate: jest.fn() } as any)
    jest
      .spyOn(cfServiceMock, 'useGetFeatureEvaluations')
      .mockReturnValue({ data: [], loading: false, mutate: jest.fn() } as any)
    jest.spyOn(cfServiceMock, 'usePatchFeature').mockReturnValue({ data: [], loading: false, mutate: jest.fn() } as any)
    jest
      .spyOn(cfServiceMock, 'useGetAllTargets')
      .mockReturnValue({ data: [], loading: false, mutate: jest.fn() } as any)
    jest
      .spyOn(cfServiceMock, 'useDeleteFeatureFlag')
      .mockReturnValue({ data: [], loading: false, mutate: jest.fn() } as any)
  })

  test('it should render rbac menu correctly', async () => {
    renderComponent()

    userEvent.click(document.querySelectorAll("[data-icon='Options']")[0])

    await waitFor(() => expect(document.getElementsByTagName('li')[0]).toHaveTextContent(/edit/i))
  })

  test('it should render edit flag modal correctly', async () => {
    renderComponent()

    userEvent.click(document.querySelectorAll("[data-icon='Options']")[0])

    await waitFor(() => expect(document.getElementsByTagName('li')[0]).toBeInTheDocument())
    userEvent.click(document.getElementsByTagName('li')[0].getElementsByTagName('a')[0])

    await waitFor(() => expect(screen.getByTestId('edit-flag-form')).toBeInTheDocument())
    expect(screen.getByTestId('edit-flag-form')).toMatchSnapshot()
  })
})
