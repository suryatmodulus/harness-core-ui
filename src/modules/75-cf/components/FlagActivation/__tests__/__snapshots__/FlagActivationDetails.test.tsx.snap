// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlagActivationDetails it should render edit flag modal correctly 1`] = `
<form
  class="FormikForm--main"
  data-testid="edit-flag-form"
>
  <div>
    <div
      class="Layout--vertical Layout--layout-spacing-large StyledProps--main editDetailsModalContainer"
    >
      <p
        class="StyledProps--font StyledProps--main"
      >
        cf.editDetails.editDetailsHeading
      </p>
      <div
        class="bp3-form-group"
        data-id="name-0"
      >
        <label
          class="bp3-label"
          for="name"
        >
          <span
            class="Tooltip--acenter"
            data-tooltip-id="flagActivationDetails_name"
          >
            name
          </span>
           
          <span
            class="bp3-text-muted"
          />
        </label>
        <div
          class="bp3-form-content"
        >
          <div
            class="bp3-input-group"
          >
            <input
              autocomplete="off"
              class="bp3-input"
              name="name"
              type="text"
              value="new flag"
            />
          </div>
        </div>
      </div>
      <div
        class="bp3-form-group"
        data-id="description-1"
      >
        <label
          class="bp3-label"
          for="description"
        >
          <span
            class="Tooltip--acenter"
            data-tooltip-id="flagActivationDetails_description"
          >
            description
          </span>
           
          <span
            class="bp3-text-muted"
          />
        </label>
        <div
          class="bp3-form-content"
        >
          <textarea
            autocomplete="off"
            class="bp3-input bp3-fill"
            maxlength="1024"
            name="description"
          />
        </div>
      </div>
      <div
        class="StyledProps--font StyledProps--main"
      >
        <div
          class="bp3-form-group"
          data-id="permanent-2"
        >
          <div
            class="bp3-form-content"
          >
            <label
              class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox checkboxEditDetails StyledProps--main StyledProps--inline"
            >
              <input
                name="permanent"
                type="checkbox"
              />
              <span
                class="bp3-control-indicator"
              />
              <span
                class="Tooltip--acenter FormikForm--checkBoxDocTooltipLabel"
                data-tooltip-id="flagActivationDetails_permanent"
              >
                cf.editDetails.permaFlag
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        class="StyledProps--font StyledProps--main"
      >
        <div
          class="bp3-divider"
        />
      </div>
      <div
        class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
      >
        <h4
          class="StyledProps--font StyledProps--font-h4 StyledProps--main StyledProps--font-variation-h5"
          level="4"
        >
          cf.gitSync.commitChanges
        </h4>
        <div
          class="StyledProps--font StyledProps--main"
        >
          <h4
            class="StyledProps--font StyledProps--font-h4 StyledProps--main StyledProps--font-variation-h5 StyledProps--margin-bottom-small StyledProps--color StyledProps--color-grey600"
            level="4"
          >
            common.gitSync.harnessFolderLabel
          </h4>
          <div
            class="Layout--horizontal Layout--layout-spacing-small StyledProps--main formRow"
          >
            <div
              class="bp3-form-group bp3-disabled"
              data-id="gitDetails.repoIdentifier-3"
            >
              <label
                class="bp3-label"
                for="gitDetails.repoIdentifier"
              >
                <span
                  class="Tooltip--acenter"
                  data-tooltip-id="flagActivationDetails_gitDetails.repoIdentifier"
                >
                  common.git.selectRepoLabel
                </span>
                 
                <span
                  class="bp3-text-muted"
                />
              </label>
              <div
                class="bp3-form-content"
              >
                <div
                  class="bp3-input-group bp3-disabled"
                >
                  <input
                    autocomplete="off"
                    class="bp3-input"
                    disabled=""
                    name="gitDetails.repoIdentifier"
                    type="text"
                    value="harnesstest"
                  />
                </div>
              </div>
            </div>
            <div
              class="bp3-form-group bp3-disabled"
              data-id="gitDetails.rootFolder-4"
            >
              <label
                class="bp3-label"
                for="gitDetails.rootFolder"
              >
                <span
                  class="Tooltip--acenter"
                  data-tooltip-id="flagActivationDetails_gitDetails.rootFolder"
                >
                  common.gitSync.harnessFolderLabel
                </span>
                 
                <span
                  class="bp3-text-muted"
                />
              </label>
              <div
                class="bp3-form-content"
              >
                <div
                  class="bp3-input-group bp3-disabled"
                >
                  <input
                    autocomplete="off"
                    class="bp3-input"
                    disabled=""
                    name="gitDetails.rootFolder"
                    type="text"
                    value="/.harness/"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="bp3-form-group bp3-disabled"
            data-id="gitDetails.filePath-5"
          >
            <label
              class="bp3-label"
              for="gitDetails.filePath"
            >
              <span
                class="Tooltip--acenter"
                data-tooltip-id="flagActivationDetails_gitDetails.filePath"
              >
                common.git.filePath
              </span>
               
              <span
                class="bp3-text-muted"
              />
            </label>
            <div
              class="bp3-form-content"
            >
              <div
                class="bp3-input-group bp3-disabled"
              >
                <input
                  autocomplete="off"
                  class="bp3-input"
                  disabled=""
                  name="gitDetails.filePath"
                  type="text"
                  value="/flags.yaml"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="StyledProps--font StyledProps--main"
        >
          <h4
            class="StyledProps--font StyledProps--font-h4 StyledProps--main StyledProps--font-variation-h5 StyledProps--margin-bottom-small StyledProps--color StyledProps--color-grey600"
            level="4"
          >
            common.gitSync.commitDetailsLabel
          </h4>
          <div
            class="bp3-form-group"
            data-id="gitDetails.commitMsg-6"
          >
            <label
              class="bp3-label"
              for="gitDetails.commitMsg"
            >
              <span
                class="Tooltip--acenter"
                data-tooltip-id="flagActivationDetails_gitDetails.commitMsg"
              >
                common.git.commitMessage
              </span>
               
              <span
                class="bp3-text-muted"
              />
            </label>
            <div
              class="bp3-form-content"
            >
              <textarea
                autocomplete="off"
                class="bp3-input bp3-fill"
                maxlength="1024"
                name="gitDetails.commitMsg"
                placeholder="common.git.commitMessage"
              />
            </div>
          </div>
          <div
            class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-flex-start StyledProps--flex-alignItems-center"
          >
            <span
              class="bp3-icon StyledProps--main"
              data-icon="git-branch-existing"
            >
              <svg
                fill="none"
                height="16"
                viewBox="0 0 16 9"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M12.403 3.67H16v1.626h-3.597a4.477 4.477 0 01-8.807 0H0V3.67h3.596a4.477 4.477 0 018.807 0zm-7.26.813a2.857 2.857 0 105.714 0 2.857 2.857 0 00-5.714 0z"
                  fill="#383946"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            <span
              class="StyledProps--font StyledProps--main StyledProps--margin-left-small StyledProps--margin-right-small StyledProps--inline"
            >
              common.git.existingBranchCommitLabel
              :
            </span>
            <div
              class="bp3-form-group bp3-disabled"
              data-id="gitDetails.branch-7"
              style="margin-bottom: 0px;"
            >
              <div
                class="bp3-form-content"
              >
                <div
                  class="bp3-input-group bp3-disabled"
                >
                  <input
                    autocomplete="off"
                    class="bp3-input"
                    disabled=""
                    name="gitDetails.branch"
                    type="text"
                    value="main"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="StyledProps--font StyledProps--main StyledProps--padding-left-xlarge StyledProps--padding-top-small"
            data-testid="commit-details-section"
          >
            <div
              class="bp3-form-group"
              data-id="autoCommit-8"
            >
              <div
                class="bp3-form-content"
              >
                <label
                  class="bp3-control bp3-checkbox bp3-large StyledProps--font Checkbox--checkbox StyledProps--main StyledProps--inline"
                >
                  <input
                    name="autoCommit"
                    type="checkbox"
                  />
                  <span
                    class="bp3-control-indicator"
                  />
                  <span
                    class="Tooltip--acenter FormikForm--checkBoxDocTooltipLabel"
                    data-tooltip-id="flagActivationDetails_autoCommit"
                  >
                    cf.gitSync.autoCommitLabel
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="StyledProps--font StyledProps--main"
      >
        <button
          class="bp3-button Button--button StyledProps--font StyledProps--main StyledProps--intent-primary Button--with-current-color"
          type="submit"
        >
          <span
            class="bp3-button-text"
          >
            save
          </span>
        </button>
        <button
          class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--without-shadow"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            cancel
          </span>
        </button>
      </div>
    </div>
  </div>
</form>
`;
